### ISSUE TITLE
[Bug]: Cross-language pipeline options are not picked up in Java DoFns

### ISSUE THREAD

**Issue by ahmedabu98** [apache/beam#33074]
### What happened?

I noticed this when testing with Managed BigQueryIO in Python SDK:

```python
options = PipelineOptions([
  "--project=google.com:clouddfe",
  "--region=us-central1",
  "--temp_location=gs://ahmedabualsaud-wordcount/tmp",
])


with beam.Pipeline(options=options) as p:
  (p
   | beam.Create([beam.Row(int=999)])
   | beam.managed.Write("bigquery",
                        config={"table": "google.com:clouddfe.ahmedabualsaud.test_table"},
                        expansion_service="localhost:12345"
                        ))
```

I would expect these pipeline options to be picked up on Java's side. With some logging, I see the Java SDK recognizing the options at expand time in PTransforms (i.e. expand()), but not at runtime in DoFns (i.e. @ProcessElement).

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Infrastructure
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

**Comment by Suvrat1629** [comment]
@ahmedabu98 I am trying to understand how this would work. So do i tweak the `expand()` method and the "@ProcessElement" to pass it to the DoFns or is there some other way to do this?

**Comment by tarun-google** [comment]
.take-issue

**Comment by tarun-google** [comment]
Posting Analysis. The issue is made of 2 sub issues. 

1. Firstly, the issue is only in the ManagedTransforms path. PipelineOptions are not passed to [PCollection input](https://github.com/apache/beam/blob/d6422d1277ad063a5991d78c25fa8fe209ad2dd4/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/providers/BigQueryFileLoadsSchemaTransformProvider.java#L93) in the managedTransforms path. But it is still passed correctly to the runner context. It is passed properly in regular cross lang path. So, this is not cross-lang issue

2. This temp_location is wrapped into [ValueProvider](https://github.com/apache/beam/blob/d6422d1277ad063a5991d78c25fa8fe209ad2dd4/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/providers/BigQueryFileLoadsSchemaTransformProvider.java#L110) following BigQueryIO [requirements.In](http://requirements.in/) later stages we have a [check](https://github.com/apache/beam/blob/d6422d1277ad063a5991d78c25fa8fe209ad2dd4/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java#L592) to read from context options if customGcsTempLocation is Null. But this was never null, this is an object we created wrapping null value. we get customGcsTempLocation non Null and customGcsTempLocation.get() null leading us not even consume from options in context

Fix
1. use customGcsTempLocation.get() for check. so we will end up reading from options in context(verified working)
2. Fix pipelineoptions dropping issue in the path to ManagedTransforms pipeline creation.

**Comment by Abacn** [comment]
This is still relevant and failing Beam 2.69.0rc3: https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2907#issuecomment-3432610269

**Comment by ahmedabu98** [comment]
> Firstly, the issue is only in the ManagedTransforms path. PipelineOptions are not passed to [PCollection input](https://github.com/apache/beam/blob/d6422d1277ad063a5991d78c25fa8fe209ad2dd4/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/providers/BigQueryFileLoadsSchemaTransformProvider.java#L93) in the managedTransforms path.

@tarun-google can you elaborate? I just ran a pipeline and printed out the options at that particular line, and I see all the options as expected

**Comment by tarun-google** [comment]
I can only reproduce this issue on Dataflow when i enable managed Transforms with  --experiments=enable_managed_transforms, which forces the ManagedTransforms upgrade. when you run this on local or with out managedTransforms upgrade step. This issue is not reproduced.

**Comment by ahmedabu98** [comment]
cc @chamikaramj  are pipeline options not included in the graph that we send to the DF service? or is the service somehow dropping the options?

**Comment by Abacn** [comment]
> I can only reproduce this issue on Dataflow when i enable managed Transforms with --experiments=enable_managed_transforms, which forces the ManagedTransforms upgrade. when you run this on local or with out managedTransforms upgrade step. This issue is not reproduced.

It could be a related issue, but the jobs in https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/2907 doesn't have managed transform upgrade enabled (no ` --experiments=enable_managed_transforms`)


EDIT: sorry, it does have managed transform enabled. Checked job log it shows

```
INFO 2025-10-22T02:29:39.824368521Z Managed transform(s) ref_AppliedPTransform_WriteToBigQuery-beam-schematransform-org-apache-beam-bigquery_write-v1_5(beam:schematransform:org.apache.beam:bigquery_write:v1) are being processed by Dataflow. This may take a few minutes.
INFO 2025-10-22T02:31:34.286652739Z I1022 02:30:31.554611 14 managed_transforms_worker_main.cc:63] Running Managed Transforms Worker main...
Successfully processed managed transform(s) ref_AppliedPTransform_WriteToBigQuery-beam-schematransform-org-apache-beam-bigquery_write-v1_5(beam:schematransform:org.apache.beam:bigquery_write:v1).
```

**Comment by tarun-google** [comment]
But still given the fix we made to allow using temp_location as backup on run-time. This should not NPE create issue in Templates as long as temp_location is passed properly. At least integration tests we fixed in Beam shows that.

**Comment by Abacn** [comment]
I may understand what happened now. Managed transform upgrade overrides the BigQueryIO from 2.69.0rc3->2.68, which did not have the fix #36209 and thus caused the issue.

@chamikaramj Although this is WAI, the implication is after Beam 2.69.0 released and before managed transform configuration changed to use Beam 2.69.0, managed BQIO will break for Beam YAML running on Beam 2.69.0

Also, is it expected for Beam YAML that not providing enable_managed_transforms experiments, managed transforms are still being processed?

**Comment by chamikaramj** [comment]
I see. Can we make sure that we release managed config before switching templates to the new version ? Also, as a workaround you can provide the pipeline option `--updateCompatibilityVersion=2.68.0` which will prevent YAML from switching to managed.

**Comment by chamikaramj** [comment]
> Also, is it expected for Beam YAML that not providing enable_managed_transforms experiments, managed transforms are still being processed?

This experiment does not have to be specified by end users and is on by default unless you run using a dev/SNAPSHOT Beam version.

Also, the underlying issue here is managed switching YAML to use BQ batch loads for batch pipelines (instead of storage write API) which requires Tarun's fix.

**Comment by Abacn** [comment]
> I see. Can we make sure that we release managed config before switching templates to the new version ? Also, as a workaround you can provide the pipeline option `--updateCompatibilityVersion=2.68.0` which will prevent YAML from switching to managed.

Confirmed that adding `--update_compatibility_version=2.68.0` pipeline option worked for 2.69.0rc3.

**Comment by ahmedabu98** [comment]
I think the original issue here is not solved right? A Java DoFn's process context does not include pipeline options passed thru the Python SDK

**Comment by Abacn** [comment]
> A Java DoFn's process context does not include pipeline options passed thru the Python SDK

I remember I checked the pipeline option is honored. The NPE in the example in description was due to other issue (https://github.com/apache/beam/issues/33074#issuecomment-3304740621)

**Comment by tarun-google** [comment]
Yes the main issue is still not fixed. As i mentioned, In this code we pass pipeline options through PCollection Input(which i think is a rare code pattern) and Runner Context. The first path is not working, we fall back to the second path.

**Comment by chamikaramj** [comment]
My understanding is that the same set of pipeline options are available to all SDKs via UW but specific options might be not available due to the differences in naming conventions. I haven't tried accessing options via DoFn's process context though and it's possible that some options get dropped when accessing them via that path.

**Comment by shunping** [comment]
@tarun-google @chamikaramj, is this a release blocker or could we move this to the next release?

**Comment by chamikaramj** [comment]
I don't think this is a release blocker.

**Comment by shunping** [comment]
Sounds good. Will move it to the next release then.

**Comment by damccorm** [comment]
Possibly relevant - https://github.com/apache/beam/pull/36443 fixed passing pipeline options from Python to Java through external transforms, which may have been the initial issue (sounds like we have more to do here)
